dtls_debug.c

```diff
@@ -241,28 +241,51 @@ dsrv_print_addr(const session_t *addr, char *buf, size_t len) {
 #endif /* NDEBUG */
 
 #ifndef WITH_CONTIKI
-void
-dsrv_log(log_t level, const char *format, ...) {
-  static char timebuf[32];
-  va_list ap;
-  FILE *log_fd;
-
-  if (maxlog < (int) level)
-    return;
 
-  log_fd = level <= DTLS_LOG_CRIT ? stderr : stdout;
+static void
+dtls_logging_handler(log_t level, const char *message) {
+  static char timebuf[32];
+  FILE* log_fd = level <= DTLS_LOG_CRIT ? stderr : stdout;
 
   if (print_timestamp(timebuf,sizeof(timebuf), time(NULL)))
     fprintf(log_fd, "%s ", timebuf);
 
   if (level <= DTLS_LOG_DEBUG)
     fprintf(log_fd, "%s ", loglevels[level]);
 
+  fwrite(message, strlen(message), 1, log_fd);
+  fflush(log_fd);
+}
+
+static dtls_log_handler_t log_handler = dtls_logging_handler;
+
+void
+dtls_set_log_handler(dtls_log_handler_t app_handler) {
+  if (log_handler == NULL)
+    log_handler = dtls_logging_handler;
+  else
+    log_handler = app_handler;
+}
+
+#ifndef DTLS_DEBUG_BUF_SIZE
+#define DTLS_DEBUG_BUF_SIZE 128
+#endif
+
+void
+dsrv_log(log_t level, const char *format, ...) {
+  va_list ap;
+  char message[DTLS_DEBUG_BUF_SIZE];
```

Comment:
Since we have `dtls_mutex.h` we could also provide a version with a static buffer here.

Reply:
Yes, will update it shortly.

Reply:
Code pushed

Reply:
I would welcome to add a short comment about calling the handler with the mutex locked.

This PR doesn't change that, but also the `static char timebuf[32];` requires such protection. Maybe that gets fixed later in an additional PR.

Reply:
Actually,` dsrv_log() `should NOT be called with 'static_mutex' locked as it does that within the function.  I agree that` timebu[32]` ought to be protected.

Reply:
My point about adding a comment is more, that if this callback locks an other mutex, and also someone calls the logging with that mutex locked, then this will be in danger of a deadlock. Therefore I think to add a comment, that informs about the mutex helps to prevent from such an deadlock.

Reply:
Added comment and code pushed.